File: Python/graphs/Articulation_Point.py
'''
Articulation Point:- In an undirected graph, any vertex whose removal
        from the graph separates/divides the graph into 2 or more graph
        which are not connected by any edge is called a cut-vertex
        or Articulation Point.

Purpose:- To return a list of all Articulation point in the given
            graph if such a point exists.
Problem Link:- https://www.geeksforgeeks.org/articulation-points-or-cut-vertices-in-a-graph

Method - Depth First Search
Intuition:- The idea behind the solution is to use DFS because it forms a DFS tree.
        In DFS tree, a vertex u is articulation point if one of the following two conditions is true.
        1) u is the root of DFS tree and it has at least two children.
        2) u is not the root of DFS tree and it has a child v such that
            no vertex in the subtree rooted with v has a back edge to one
            of the ancestors (in DFS tree) of u.

Time Complexity:-  O(N+M)
Space Complexity:  O(N)

Argument: Dictionary ( Graph vertex numbered 1 to N)
Return  : List       ( Articulation Point)

'''
from collections import defaultdict

# Global Variable to keep a track of finishing times
Time = 0


def dfs(graph, node, parent, low, f_time, visited, AP):
    global Time

    # Mark the node visited and Initialize the finishing time
    visited[node] = True
    low[node] = Time
    f_time[node] = Time
    Time += 1

    # To count the children from the current Node
    child = 0

    # Check for each adjoining Vertex/Node
    for i in graph[node]:

        if visited[i]:
            low[node] = min(low[node], f_time[i])

        if not visited[i]:
            parent[i] = node
            child += 1
            dfs(graph, i, parent, low, f_time, visited, AP)
            low[node] = min(low[node], low[i])

            # (1) u is the root of the DFS tree and has two or more children.
            if parent[node] == -1 and child > 1:
                AP[node] = 1

            # (2) If u is not root and low value of one of its child is more
            # than discovery value of u.
            if f_time[node] <= low[i] and parent[node] != -1:
                AP[node] = 1


def Articulation_Point(n, graph):

    # Initialize parent finishing time and Low
    parent = [-1] * (n + 1)
    low = [float('inf') for i in range(n + 1)]
    f_time = [float('inf') for i in range(n + 1)]

    # To keep a check on already visited nodes
    visited = [0] * (n + 1)

    # To store the Articulation point
    AP = [0] * (n + 1)

    # Call the DFS function to recursively check for each of the vertex
    for i in graph.keys():
        if(not visited[i]):
            dfs(graph, i, parent, low, f_time, visited, AP)

    # Collect all the Articulation point in a list
    ans = []
    for i in range(n + 1):
        if AP[i]:
            ans += [i]

    # Return all the Articulation Point
    return ans


# -------------------------------Driver Code-------------------------------

if __name__ == "__main__":
    n, m = map(int, input("Enter the number of vertex and edges: ").split())
    print("Enter the edges: ")
    graph = defaultdict(list)

    for i in range(m):
        a, b = map(int, input().split())
        graph[a] += [b]
        graph[b] += [a]
    ans = Articulation_Point(n, graph)

    # Print the Answer
    if not ans:
        print("There is No Articulation Point in the Graph")
    else:
        print("The Articulation point in the given graph are : ", *ans)

"""
Sample Input / Output

     5----1------2
     \     \
      \     \
       \     \
 3------6-----4

Enter the number of vertex and edges: 6 6
Enter the edges:
5 1
3 6
6 4
1 2
5 6
1 4
The Articulation point in the given graph are :  1 6

     5----1------2
    /\     \    /
   /  \     \  /
  /    \     \/
 3------6-----4

Enter the number of vertex and edges: 6 8
Enter the edges:
5 1
1 2
1 4
5 6
6 4
4 2
3 6
5 3
There is No Articulation Point in the Graph
"""

File: Java/graphs/Articulation_Point.java
/*Problem Description
A vertex is said to be an articulation point in a graph if removal of the vertex and associated edges disconnects the graph.
You are given the nodes and edges . Find the articulation point.
*/

import java.util.*;
import java.lang.*;
import java.io.*;

public class Articulation_Point {
        // Initialize all the required variables with their type. 
	private static int[] low =  new int[10001];
	private static int[] visited = new int[10001]; 
	private static int[] disc = new int[10001] ;
	private static int timer = 1;
	private static Set<Integer> hashset ;
        static ArrayList<Integer> adj[];
        //  This class will help us to create a graph and know the articulation point.
	static class Graph
	 {
		 static int v ;		
       		 // Creating graph 
		 Graph(int v)
		 {
			 Graph.v = v;
			 adj = new ArrayList[v+1];			 
			 for(int i = 0 ; i < v+1 ; i++)
			 {
				 adj[i] = new ArrayList<Integer>();
			 }
		 }		 
     	         // adding nodes with each other 
		 static void addEdge(int u , int v)
		 {
			 adj[u].add(v);
			 adj[v].add(u);
		 }
		// The method will help us to know the articulation point
		/*
		    Code flow : 
		    step 1 :  We need to calculate the lowest discovery number. 
		    step 2 : The algorithm then searches for its adjacent vertices.
		    step 3 : the algorithms take a pair of vertices and checks whether its an articulation point or not.
		    step 4 : There’s one special case when the root is an articulation point
		*/    
		 static void dfs(int node , int par)
			{
				visited[node] = 1 ;
				low[node] = disc[node] = timer++;
				int children = 0 ;
				for(int child : adj[node] )
				{
					if(child == par) continue;
					if(visited[child] == 0 )
					{
						children++;
						dfs(child , node );
						if(disc[node] <= low[child] && par != -1)
						{
							hashset.add(node );
						}
						low[node] = Math.min(low[node] , low[child] );
					}
					else
					{
						low[node] = Math.min(low[node], disc[child] );
					}									
				}
              		       // if root is articulation point
				if(par == 0  && children > 1)
				{
					hashset.add(node);
				}			
			}
	 }
	public static void main(String[] args) throws IOException 
	{ 
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter number of testcase");
		int test  = sc.nextInt();
		while(test-- > 0)
		{			
           		 // Clear all the values inside the array to start new fresh array for next test case
			 Arrays.fill(disc, 0);
			 Arrays.fill(visited, 0);
			 Arrays.fill(low, 0);
			 timer = 0 ;
            		 // hashset for storing the articulation point
			 hashset = new HashSet<Integer>();
             		 // No of nodes
			 System.out.println("Enter number of nodes");
			 int n = sc.nextInt();
           		 // No of edges  
			 System.out.println("Enter number of edges");
			 int m = sc.nextInt();
           		 // Create a graph with 10001 entries so it does not overflow
			 Graph g = new Graph(10001);  
			 // Add the nodes with each other
			 System.out.println("Enter values of edges");
			for(int i  = 1 ; i <= m ; i++)
			{
				int a = sc.nextInt();
				int b = sc.nextInt();
				g.addEdge(a,b);
			}
			// visit every element of nodes and accordingly call dfs function 			
			for(int i = 1 ; i < n + 1 ; i++ )
			{
				if(visited[i] == 0 )
					g.dfs(i , -1 );
			}		
			// Print all the articulation point
			for(Integer val : hashset)
			{
				System.out.print(val + " ");
			}	 
		}
  }	
}

/* Input
Enter number of testcase
1
Enter number of nodes
4
Enter number of edges
5
Enter values of edges
1 0
0 2
2 1
0 3
3 4

Output
0 3

Explanation :
From all the nodes we find that articulation point is 0 and 3.This will seperate the graph and can make many unconnect graph
Time Complexity : O(V + E)
Space Complexity : O(V + E)
*/
